### PR TITLE
Use Theano's abstract interface for batch normalization

### PR DESCRIPTION
**Pull request by gvtulder**
This is a continuation of #4162. The patches change the batch normalization functions to use Theano's new abstract batch normalization functions, once they are included (see Theano/Theano#5190).

The current version works for me with the appropriate Theano branch. I'll need to have another look at simplifying the code further, though. I think the parameter dimshuffle can be removed, since it should be handled by the Theano functions.

### CODE REVIEW

**Comment by fchollet** [comment]
Can you provide a summary of the changes? There seems to be several things going on in this PR.

**Comment by gvtulder** [comment]
I've reorganized things a bit so the sequence of patches might make more sense now. (And I changed a bit in the code as well, so it's slightly different from the previous iteration.)

The first three patches do a bit of refactoring:

1. Remove a bit of duplicate code in the BatchNormalization layer.

2. Introduce a `K.normalize_batch_in_testing` function in the backends. This makes the testing function more similar to `K.normalize_batch_in_training`, because it accepts a list of `reduction_axes` and will broadcast/reshape/dimshuffle the parameters. This is useful for the following patches.
This patch also deprecates the existing `K.batch_normalization` function, which might be useful to keep around for external code that depends on this.

3. Use the new `K.normalize_batch_in_testing` in the BatchNormalization layer, to replace the old call to `K.batch_normalization`. This simplifies the code because the broadcast/reshape/dimshuffle is handled by the backend function. This patch also removes the unsightly `if K.backend() == 'tensorflow'` test.

The fourth patch starts using the new Theano interface:

4. The code in `K.normalize_batch_in_training` and `K.normalize_batch_in_testing` can be replaced with shorter calls to the new Theano functions. The old implementation is still there for compatibility with older Theano versions.

Finally, there is a small issue with `epsilon` in the new Theano interface. Perhaps the Theano patch will be changed to handle this in a nicer way, but at the moment `epsilon=1e-5` is an unlucky default value. When 1e-5 is converted to float32 the value is actually smaller than 1e-5 in float64, which will cause Theano to reject it.

5. This patch changes the default value for epsilon to something slightly larger than 1e-5. It's not ideal, but at the moment I can't think of a better solution.

**Comment by fchollet** [comment]
Let's change the epsilon default to 1e-3. Let's also use the same default in the backend.

The current BN code can definitely be refactored for simplicity, but I don't think having `normalize_batch_in_training` and `normalize_batch_in_testing` is the right way to do it. If anything we should remove the former instead of adding the latter.

Let's try to keep the API close to the TF API, in particular by having `batch_normalization` with the same API as TF.

**Comment by fchollet** [comment]
I also don't think that we should have different reshaping (broadcasting) behavior across the two backends. The reason is that it would tie BN layer weights to a specific backend: you couldn't load some Theano-trained models in TF.

Having the same behavior fixes this, and also allows us to move the broadcasting code to the BN layer's `call`.

**Comment by gvtulder** [comment]
Good point, I hadn't really looked at how it works in TensorFlow. But allow me to make a few counterpoints.

> Let's try to keep the API close to the TF API, in particular by having batch_normalization with the same API as TF.

I do think that it is useful to have a separate method for batch normalization during training, since there are also separate training and testing functions in cuDNN that you might want to use. That's why there are separate methods in Theano. In TensorFlow the function for both is called `tf.fused_batch_norm`, I think, with a parameter to specify the training/testing.

As an aside: I don't know enough about TensorFlow, but wouldn't it might be more efficient to use `tf.fused_batch_norm` in the TF backend anyway? It looks like `tf.batch_normalization` won't use cuDNN.

> I also don't think that we should have different reshaping (broadcasting) behavior across the two backends. The reason is that it would tie BN layer weights to a specific backend: you couldn't load some Theano-trained models in TF.

Fair point, but I don't really see how the reshaping would be different. The implementation may be different, but the behavior should be exactly the same. The parameter variables still have the same shape in both backends (they're vectors). The only difference would that the reshaping/broadcasting happens inside the backend function instead of in BatchNormalization's `call`.

And apparently you *do* want some backend-specific broadcasting, given [these lines](https://github.com/fchollet/keras/blob/c4c4fac1aed6ccc84d2f21028ebdfd03213709aa/keras/layers/normalization.py#L131-L135) in BatchNormalization.call:
```python
    if K.backend() == 'tensorflow' and sorted(reduction_axes) == range(K.ndim(x))[:-1]:
        # calling batch_normalization without broadcasting
    else:
        # calling batch_normalization with broadcasted parameters
```
Moving the broadcasting to the backend functions makes it possible to implement this backend-specific optimization for TensorFlow and to let Theano handle all of the broadcasting and unbroadcasting on its own. You also don't have to deduce the `reduction_axes` list from the broadcasting pattern if you simply pass this list to the backend function.

> Having the same behavior fixes this, and also allows us to move the broadcasting code to the BN layer's call.

Why would you prefer to keep the broadcasting in BN's `call`?

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -727,11 +727,41 @@ def normalize_batch_in_training(x, gamma, beta,
     return normed, mean, var
 
 
+def normalize_batch_in_testing(x, mean, var, beta, gamma,
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I think we should rename this to `batch_normalization` in order to follow the TF API. API consistency matters, and Keras is aligning on TF rather than Theano.

File: keras/layers/normalization.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ class BatchNormalization(Layer):
     # References
         - [Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](http://jmlr.org/proceedings/papers/v37/ioffe15.pdf)
     '''
-    def __init__(self, epsilon=1e-5, mode=0, axis=-1, momentum=0.99,
+    def __init__(self, epsilon=1.001e-5, mode=0, axis=-1, momentum=0.99,
```
Comment (fchollet) [inline keras/layers/normalization.py]: Let's change epsilon to 1e-3.

**Comment by gvtulder** [comment]
Okay, revised version:

* There are now no changes to the TensorFlow backend.
* I changed the Theano backend to handle the case with vector mean, var etc. This should now give the same result as the TensorFlow backend, which means that the ugly backend-specific case in the BatchNormalization layer can be removed.
* There is more reshaping than strictly required for Theano, but most of this should be optimized out while compiling.
* The default epsilon is now 1e-3.

Of course, all this still depends on the Theano PR that should introduce the new interface.

**Comment by fchollet** [comment]
LGTM, thanks!

**Comment by gvtulder** [comment]
@fchollet: Thanks for merging, but I probably should have phrased the warning at the end of my previous post a bit stronger. Merging this is a bit premature, because the new interface in Theano is not yet available and therefore a moving target. The Theano pull request has not yet been merged (see Theano/Theano#5190), so it's still possible that the interface changes a bit. For now it's not a big problem because Keras will just use the current version, but I might have to come back with a new PR if the Theano interface does change before it is merged. (I don't know if it will. I obviously think it's very good the way it is now, but it hasn't been thoroughly reviewed.) My apologies for the confusion.

**Comment by fchollet** [comment]
The current is backwards-compatible, so this should not be an issue, right?

**Comment by gvtulder** [comment]
Yes, it's fully backwards-compatible.
There is a small chance that it will not be forward-compatible. This will only be an issue if the new Theano functions get a different implementation or require different parameters, while keeping the function names I have given them now. I guess that is not very likely to happen.

**Comment by fchollet** [comment]
Lack of forward compatibility is simply "a bug". This should be detected by our unit tests, and quickly fixed if it happens.

I'd rather merge PRs that are approved ASAP, if that's possible, instead of having them wait indefinitely.
